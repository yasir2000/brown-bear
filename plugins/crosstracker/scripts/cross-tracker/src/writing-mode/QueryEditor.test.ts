/*
 * Copyright (c) Enalean, 2018-Present. All Rights Reserved.
 *
 * This file is a part of Tuleap.
 *
 * Tuleap is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Tuleap is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Tuleap. If not, see <http://www.gnu.org/licenses/>.
 */

import type { Wrapper } from "@vue/test-utils";
import { shallowMount } from "@vue/test-utils";
import { createCrossTrackerLocalVue } from "../helpers/local-vue-for-test";
import QueryEditor from "./QueryEditor.vue";
import WritingCrossTrackerReport from "./writing-cross-tracker-report";

describe("QueryEditor", () => {
    async function instantiateComponent(
        writingCrossTrackerReport: WritingCrossTrackerReport
    ): Promise<Wrapper<QueryEditor>> {
        return shallowMount(QueryEditor, {
            propsData: {
                writingCrossTrackerReport,
            },
            localVue: await createCrossTrackerLocalVue(),
        });
    }

    it("Displays a code mirror integration", async () => {
        const writingCrossTrackerReport = new WritingCrossTrackerReport();
        writingCrossTrackerReport.expert_query = "@title = 'foo'";

        const wrapper = await instantiateComponent(writingCrossTrackerReport);
        expect(wrapper.element).toMatchSnapshot();
        expect(wrapper.vm.$data.value).toContain(writingCrossTrackerReport.expert_query);
    });

    it("Update the report chen query is updated", async () => {
        // eslint-disable-next-line
        document.createRange = (): any => {
            return {
                getBoundingClientRect: (): void => {
                    // nothing
                },
                setEnd: (): void => {
                    // nothing
                },
                setStart: (): void => {
                    // nothing
                },
                getClientRects(): { length: number } {
                    return { length: 0 };
                },
            };
        };

        const writingCrossTrackerReport = new WritingCrossTrackerReport();
        writingCrossTrackerReport.expert_query = "@title = 'foo'";
        const wrapper = await instantiateComponent(writingCrossTrackerReport);

        wrapper.vm.$data.code_mirror_instance.setValue("@title = 'bar'");
        expect(writingCrossTrackerReport.expert_query).toContain("@title = 'bar'");
    });
});
