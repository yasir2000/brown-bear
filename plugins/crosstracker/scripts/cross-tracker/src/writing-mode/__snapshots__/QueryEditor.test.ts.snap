// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QueryEditor Displays a code mirror integration 1`] = `
<div
  class="cross-tracker-expert-content"
>
  <div
    class="cross-tracker-expert-content-query tlp-form-element"
  >
    <label
      class="tlp-label"
      data-current-language="en_US"
      data-msgid="Query"
      for="expert-query-textarea"
    >
      Query
    </label>
     
    <textarea
      class="cross-tracker-expert-content-query-textarea tlp-textarea"
      data-test="expert-query-textarea"
      id="expert-query-textarea"
      name="expert_query"
      placeholder="Example: @title = 'value'"
      style="display: none;"
      type="text"
    />
    <div
      class="CodeMirror cm-s-default CodeMirror-wrap"
    >
      <div
        style="overflow: hidden; position: relative; width: 3px; height: 0px;"
      >
        <textarea
          autocapitalize="off"
          autocorrect="off"
          spellcheck="false"
          style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;"
          tabindex="0"
        />
      </div>
      <div
        class="CodeMirror-vscrollbar"
        cm-not-content="true"
        style="display: block; bottom: 0px;"
        tabindex="-1"
      >
        <div
          style="min-width: 1px; height: 53px;"
        />
      </div>
      <div
        class="CodeMirror-hscrollbar"
        cm-not-content="true"
        tabindex="-1"
      >
        <div
          style="height: 100%; min-height: 1px; width: 0px;"
        />
      </div>
      <div
        class="CodeMirror-scrollbar-filler"
        cm-not-content="true"
      />
      <div
        class="CodeMirror-gutter-filler"
        cm-not-content="true"
      />
      <div
        class="CodeMirror-scroll"
        tabindex="-1"
      >
        <div
          class="CodeMirror-sizer"
          style="margin-left: 0px; padding-right: 0px; padding-bottom: 0px;"
        >
          <div
            style="position: relative;"
          >
            <div
              class="CodeMirror-lines"
              role="presentation"
            >
              <div
                role="presentation"
                style="position: relative; outline: none;"
              >
                <div
                  class="CodeMirror-measure"
                >
                  <pre
                    class="CodeMirror-line-like"
                  >
                    <span>
                      xxxxxxxxxx
                    </span>
                  </pre>
                </div>
                <div
                  class="CodeMirror-measure"
                />
                <div
                  style="position: relative; z-index: 1;"
                />
                <div
                  class="CodeMirror-cursors"
                />
                <div
                  class="CodeMirror-code"
                  role="presentation"
                />
              </div>
            </div>
          </div>
        </div>
        <div
          style="position: absolute; height: 50px; width: 1px; border-bottom: 0px solid transparent;"
        />
        <div
          class="CodeMirror-gutters"
          style="display: none;"
        />
      </div>
    </div>
     
    <p
      class="tlp-text-info"
    >
      <i
        class="fa fa-info-circle"
      />
       
      <translate-stub
        tag="span"
      >
        
                You can use: AND, OR, parenthesis. Autocomplete is activated with Ctrl + Space.
            
      </translate-stub>
    </p>
  </div>
   
  <div
    class="tlp-form-element"
  >
    <label
      class="tlp-label"
      data-current-language="en_US"
      data-msgid="
            Allowed fields
        "
      for="expert-query-allowed-fields"
    >
      
            Allowed fields
        
    </label>
     
    <select
      class="cross-tracker-expert-content-query-allowed-fields tlp-select"
      id="expert-query-allowed-fields"
      multiple="multiple"
      name="allowed-fields"
    >
      <option
        value="@title"
      >
        Title
      </option>
       
      <option
        value="@description"
      >
        Description
      </option>
       
      <option
        value="@status"
      >
        Status
      </option>
       
      <option
        value="@submitted_on"
      >
        Submitted on
      </option>
       
      <option
        value="@last_update_date"
      >
        Last update date
      </option>
       
      <option
        value="@submitted_by"
      >
        Submitted by
      </option>
       
      <option
        value="@last_update_by"
      >
        Last update by
      </option>
       
      <option
        value="@assigned_to"
      >
        Assigned to
      </option>
    </select>
  </div>
</div>
`;
